rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,209 @@
+4. "Read/Write Automatic" or "Two-Way" mode: originally not thought possible, but is doable via change tracking with an audit log: check the differences between the cloud state and database state, then look up via the audit log if the noted change was explicitly triggered by the user, or is from the cloud, and then decide what operation to perform, syncing changes from the cloud (which could be due to users still managing part of their infrastructure with IaC tools) and applying changes to the cloud where that was the user's intention.
+
+Each of these different "modes" have their own use-cases with various pros and cons, and they all seemed to be incompatible with each other. If the database is in one of the three automatic modes, there's no way to preview the changes that are going to happen. If the database is in read/write automatic mode, it won't restore critical infrastructure actually managed by IaSQL like the write-only automatic mode would, but write-only automatic mode prevents *any* other IaC tool from being used, so adoption would have to be all-or-nothing if that automatic restoration property was what brought someone to IaSQL in the first place.
+
+Recent work on the AWS `codebuild`, `codedeploy`, and `codepipeline` services has demonstrated that only "Two-Way" mode is actually viable: `codebuild`, for instance, accepts an arbitrary, user-defined YAML file describing the process to build code from some place and store it in some other place. IaSQL cannot know before it triggers `codebuild` what it will do, and so during the `iasql_apply` call suddenly there will exist resources, such as a new ECR image, that are not in the database which `iasql_apply` will want to remove because that's not part of the "desired" state as far as the engine is concerned, but *is* the true desired outcome from the user.
+
+Attempting to `iasql_sync` in the middle of that `iasql_apply` could cause IaSQL to drop other operations it is supposed to be performing, freezing their cloud into some unexpected intermediate state. And it is impossible to "parse" the YAML to determine what resources are going to be created. Since the YAML allows arbitrary shell script execution, the script could do something like `if (Math.random() > 0.5) { writeToS3(); } else { writeToECR(); }` or even `curl https://raw.githubusercontent.com/company/project/some_binary_that_calls_aws_apis; chmod +x some_binary_that_calls_aws_apis; ./some_binary_that_calls_aws_apis;` and there's no good, general purpose way to know what the side-effects of `codebuild` will be.
```

Comment:
lol the examples here were amongst the top 3% programming jokes I've heard. +1 for the subsequent `chmod +x` and running the script and not leaving a non-fully-runnable script.